Kinvey.Core/Core/AbstractKinveyClientRequest.cs

```diff
@@ -491,67 +491,80 @@ public async Task<HttpResponseMessage> ExecuteUnparsedAsync()
+                    Credential cred = Client.Store.Load(Client.ActiveUser.Id, Client.SSOGroupKey);
+                    string refreshToken = null;
+                    string redirectUri = null;
                     string micClientId = null;
 
-					if (cred != null)
-					{
-						refreshToken = cred.RefreshToken;
-						redirectUri = cred.RedirectUri;
+                    if (cred != null)
+                    {
+                        refreshToken = cred.RefreshToken;
+                        redirectUri = cred.RedirectUri;
                         micClientId = cred.MICClientID;
 
-						if (!string.IsNullOrEmpty(refreshToken) && !refreshToken.ToLower().Equals("null"))
-						{
-							//use the refresh token for a new access token
-                            JObject result = await Client.ActiveUser.UseRefreshToken(refreshToken, redirectUri, micClientId).ExecuteAsync();
-
-							// log out the current user without removing the user record from the credential store
-							Client.ActiveUser.LogoutSoft();
-
-							//login with the access token
-							Provider provider = new Provider();
-							provider.kinveyAuth = new MICCredential(result["access_token"].ToString());
-							User u = await User.LoginAsync(new ThirdPartyIdentity(provider), Client);
-
-							//store the new refresh token
-							Credential currentCred = Client.Store.Load(Client.ActiveUser.Id, Client.SSOGroupKey);
-							currentCred.AccessToken = result["access_token"].ToString();
-                            currentCred.RefreshToken = result.GetValidValue("refresh_token");
-                            currentCred.RedirectUri = redirectUri;
-							Client.Store.Store(Client.ActiveUser.Id, Client.SSOGroupKey, currentCred);
-
-							// Retry the original request
-							RequestAuth = new KinveyAuthenticator(currentCred.AuthToken);
-							var retryResponse = await ExecuteUnparsedAsync();
-							return retryResponse;
-						}
-						else
-						{
-							//logout the current user
-							Client.ActiveUser.Logout(); // TODO is this a potential deadlock?
-						}
-					}
-					else
-					{
-						Client.ActiveUser.Logout();
-					}
-				}
-				else
-				{
-					Client.ActiveUser.Logout();
-				}
-			}
+                        if (!string.IsNullOrEmpty(refreshToken) && !refreshToken.ToLower().Equals("null"))
+                        {
+                            if (!hasRetried)
+                            {
+                                // Attempting retry - set flag to prevent additional attempts
+                                hasRetried = true;
+
+                                //use the refresh token for a new access token
+                                JObject result = await Client.ActiveUser.UseRefreshToken(refreshToken, redirectUri, micClientId).ExecuteAsync();
+
+                                // log out the current user without removing the user record from the credential store
+                                Client.ActiveUser.LogoutSoft();
```

Comment:
@vinaygahlawat I don't think we have this method in other platforms. Should we implement it across all SDKs?

Reply:
@heyzooi Yes, I think we should have a method like `LogoutSoft()` in all SDKs, specifically for handling the refresh token case.